controllers/storageclusterpeer/storageclusterpeer_controller.go

```diff
@@ -18,46 +18,160 @@ package storageclusterpeer
+		return reconcile.Result{}, err
+	}
+
+	result, reconcileErr := r.reconcilePhases()
+
+	// Apply status changes to the StorageClient
+	statusErr := r.Client.Status().Update(ctx, r.storageClusterPeer)
+	if statusErr != nil {
+		r.log.Error(statusErr, "Failed to update StorageClusterPeer status.")
+	}
+
+	if reconcileErr != nil {
+		err = reconcileErr
+	} else if statusErr != nil {
+		err = statusErr
+	}
+
+	return result, err
+}
+
+func (r *StorageClusterPeerReconciler) reconcilePhases() (ctrl.Result, error) {
+	ocsClient, err := r.newExternalClusterClient()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	defer ocsClient.Close()
+
+	// marked for deletion
+	if !r.storageClusterPeer.GetDeletionTimestamp().IsZero() {
+		//TODO: Removing PeerOCS Call
+
+		if controllerutil.RemoveFinalizer(r.storageClusterPeer, storageClusterPeerFinalizer) {
+			r.log.Info("removing finalizer from StorageClusterPeer.", "StorageClusterPeer", r.storageClusterPeer.Name)
+			if err := r.update(r.storageClusterPeer); err != nil {
+				r.log.Info("Failed to remove finalizer from StorageClusterPeer", "StorageClusterPeer", r.storageClusterPeer.Name)
+				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClient: %v", err)
+			}
+		}
+	}
+
+	if controllerutil.AddFinalizer(r.storageClusterPeer, storageClusterPeerFinalizer) {
+		r.log.Info("Finalizer not found for StorageClusterPeer. Adding finalizer.", "StorageClusterPeer", r.storageClusterPeer.Name)
+		if err := r.update(r.storageClusterPeer); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to update StorageClusterPeer: %v", err)
+		}
+	}
+
+	if r.storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		return r.peerOCS(ocsClient)
+	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *StorageClusterPeerReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1.StorageClusterPeer{}).
-		Complete(r)
+func (r *StorageClusterPeerReconciler) newExternalClusterClient() (*providerClient.OCSProviderClient, error) {
+
+	ocsProviderClient, err := providerClient.NewProviderClient(
+		r.ctx, r.storageClusterPeer.Spec.ApiEndpoint, time.Second*10)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+
+	return ocsProviderClient, nil
+}
+
+func (r *StorageClusterPeerReconciler) peerOCS(ocsClient *providerClient.OCSProviderClient) (ctrl.Result, error) {
+	r.storageClusterPeer.Status.State = ocsv1.StorageClusterPeerStatePending
+	response, err := ocsClient.PeerStorageCluster(r.ctx, r.storageClusterPeer.Spec.OnboardingToken, string(r.storageCluster.UID))
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("failed to Peer Storage Cluster, code: %v.", status.Code(err)))
+		return ctrl.Result{}, err
+	}
+	r.storageClusterPeer.Status.PeerInfo.StorageClusterUid = response.StorageClusterUID
```

Comment:
PeerInfo should be a pointer. If peering is not completed there should not be a substruct on the API status.
With that in mind, the above line has the potential to throw a nil access error.